src/__tests__/reactUtils-test.js

```diff
@@ -0,0 +1,232 @@
+
+describe('useHasStayedTrueForMs', () => {
+  /**
+   * Simulate a mock component using the hook, and inspect hook's value.
+   *
+   * - The constructor mounts the component. (`ms` won't change through the
+   *   lifetime of the TestMachine instance.)
+   * - Use `updateValue` to change the `value` arg passed to the hook.
+   * - Use `hookOutput` to get the return value of the hook from the latest
+   *   render.
+   * - (Important) To wait for a duration, use the instance's `sleep` method
+   *   instead of the util `fakeSleep` or similar. It wraps some
+   *   `react-test-renderer` boilerplate.
+   * - When done, call 'cleanup'.
+   *
+   * Encapsulates a few things:
+   * - react-test-renderer is untyped (so far)
+   * - boilerplate for using react-test-renderer, like calling `act`
+   *   repeatedly
+   * - boring details like how the mock component is implemented
+   */
+  // I'm not totally clear on everything `act` does, but react-test-renderer
+  // seems to recommend it strongly enough that we actually get errors if we
+  // don't use it. Following links --
+  //   https://reactjs.org/docs/test-renderer.html#testrendereract
+  //   https://reactjs.org/docs/test-utils.html#act
+  //   https://reactjs.org/docs/testing-recipes.html
+  // -- I see the following, which I think does the best job of explaining.
+  // (The `act` in `react-dom/test-utils` might not be identical to the
+  // `act` in `react-test-renderer`, but `react-test-renderer` says they're
+  // similar.)
+  // > When writing UI tests, tasks like rendering, user events, or data
+  // > fetching can be considered as “units” of interaction with a user
+  // > interface. `react-dom/test-utils` provides a helper called `act()`
+  // > that makes sure all updates related to these “units” have been
+  // > processed and applied to the DOM before you make any assertions
+  class TestMachine {
+    static HOOK_VALUE_TRUE = 'HOOK_VALUE_TRUE';
+    static HOOK_VALUE_FALSE = 'HOOK_VALUE_FALSE';
+
+    _TestComponent: ComponentType<{| value: boolean |}>;
+    _testRenderer: $FlowFixMe;
+
+    constructor(ms: number, initialValue: boolean) {
+      this._TestComponent = function _TestComponent(props: {| value: boolean |}) {
+        const hookOutput = useHasStayedTrueForMs(props.value, ms);
+        return hookOutput ? TestMachine.HOOK_VALUE_TRUE : TestMachine.HOOK_VALUE_FALSE;
+      };
+      this._testRenderer = this._createTestRenderer(initialValue);
+    }
+
+    updateValue(value: boolean) {
+      this._updateTestRenderer(value);
+    }
+
+    hookOutput() {
+      const result = this._testRenderer.root.children[0] === TestMachine.HOOK_VALUE_TRUE;
+      return result;
+    }
+
+    // eslint-disable-next-line class-methods-use-this
+    async sleep(ms: number): Promise<void> {
+      // The hook uses `useState`, which seems to make `react-test-renderer`
+      // complain if we don't use `act`.
+      return act(() => fakeSleep(ms));
```

Comment:
Yeah. Here's the error if I remove the `act`:
```
  console.error
    Warning: An update to _TestComponent inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
    
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
        at _TestComponent (/home/greg/z/mobile/src/__tests__/reactUtils-test.js:54:71)

      70 |   useEffect(() => {
      71 |     if (value) {
    > 72 |       const id = setTimeout(() => setResult(true), duration);
```

I think the point is that `fakeSleep` causes that timer to run -- that's why we're calling it, after all -- which means we set the state. That causes an update to the test component, so we need it wrapped in `act` to make sure the updates get applied.

Reply:
Yeah. It occurs to me after re-reading that error you quoted: I think they did a pretty good job raising this error and choosing its error message. It happens early, it's specific, and it's actionable.